### PR TITLE
parser: implement Restore for Constraint

### PR DESCRIPTION
**Pull request by feloxx**
1. implement Restore for Constraint

---

Issue: [https://github.com/pingcap/tidb/issues/8532](https://github.com/pingcap/tidb/issues/8532)

### CODE REVIEW

File: ast/ddl_test.go [hunk 1]
```diff
@@ -143,6 +143,23 @@ func (ts *testDDLSuite) TestDDLReferenceDefRestore(c *C) {
 	RunNodeRestoreTest(c, testCases, "CREATE TABLE child (id INT, parent_id INT, INDEX par_ind (parent_id), FOREIGN KEY (parent_id) %s)", extractNodeFunc)
 }
 
+func (ts *testDDLSuite) TestDDLConstraintRestore(c *C) {
+	testCases := []NodeRestoreTestCase{
+		{"INDEX par_ind (parent_id)", "INDEX `par_ind`(`parent_id`)"},
```
Comment (leoppro) [inline ast/ddl_test.go]: add test for `UNIQUE KEY`  `UNIQUE INDEX` `FULLTEXT`

Comment (leoppro) [reply ast/ddl_test.go]: add test for `ConstraintNoConstraint`

File: ast/ddl.go [hunk 2]
```diff
@@ -473,7 +473,58 @@ type Constraint struct {
 
 // Restore implements Node interface.
 func (n *Constraint) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	if n.Tp != ConstraintNoConstraint {
```
Comment (leoppro) [inline ast/ddl.go]: if n.Tp == ConstraintNoConstraint {
    return nil
}

**Comment by feloxx** [comment]
Add test cases;

ConstraintElem yacc `FULLTEXT` logic
update KeyOrIndex => KeyOrIndexOpt

File: ast/ddl.go [hunk 3]
```diff
@@ -473,25 +473,26 @@ type Constraint struct {
...
-			ctx.WriteKeyWord("UNIQUE KEY")
-		case ConstraintUniqIndex:
-			ctx.WriteKeyWord("UNIQUE INDEX")
-		case ConstraintForeignKey:
-			ctx.WriteKeyWord("FOREIGN KEY")
-		case ConstraintFulltext:
-			ctx.WriteKeyWord("FULLTEXT")
-		}
+	if n.Tp == ConstraintNoConstraint {
+		return nil
```
Comment (kennytm) [inline ast/ddl.go]: You could put this entirely inside the switch...

```go
switch n.Tp {
case ConstraintNoConstraint:
    return nil
...
```

**Comment by leoppro** [comment]
LGTM, waiting for @kennytm 's approving.

File: ast/ddl.go (hunk 3 above)
Comment (feloxx) [reply ast/ddl.go]: Eye of Monkey King!

**Pull request merged**